### ISSUE TITLE
Support GNU toolchain and Intel MPI

### ISSUE THREAD

**Issue by hfp** [cp2k/dbcsr#253]
See closed PR #249.

**Comment by alazzaro** [comment]
- [ ] Refactor MPI wraps (see for example https://github.com/dftbplus/mpifx)
- [ ] Remove F77 interface
- [ ] Use F08 interface

**Comment by oschuett** [comment]
I really like your idea of relying on the C API instead of dealing with external `.mod` files.

**Comment by hfp** [comment]
Of course, this issue is of similar interest for CP2K which invented the `__HAS_NO_MPI_MOD` anyway. However, with CP2K and DBCSR/Makefile all paths are implemented wrt `__HAS_NO_MPI_MOD`. The purpose of #249 was just to bring DBCSR/CMake to the same level. Still, I think CP2K does not document the combination Intel MPI with GNU Compiler properly (it requires mentioned `-D__HAS_NO_MPI_MOD` in the ARCH file).

**Comment by oschuett** [comment]
Yes, CP2K should adopt the same solution. It's probably not worth creating a separate library for this. Hence, we should just copy improvements from `dbcsr_mpiwrap.F` to CP2K's `message_passing.F`.

**Comment by vondele** [comment]
useful reading...
https://www.mcs.anl.gov/~balaji/pubs/2014/eurompi/eurompi14.f08binding.pdf
